### PR TITLE
Bump minimum Doctrine ORM version to v2.14 and add support for Doctrine ORM v3 & Lexer v3

### PR DESCRIPTION
**Pull request by GwendolenLynch**
This is a follow-up for #177

* Update `composer.json` to (re)allow Doctrine ORM v3 & Lexer v3
* Use `Doctrine\ORM\Query\TokenType` over deprecated/removed lexer constants
* Migrate entity fixtures to attributes

### CODE REVIEW

File: fixtures/MartinGeorgiev/Doctrine/Entity/ContainsIntegers.php [hunk 1]
```diff
@@ -6,23 +6,15 @@
 
 use Doctrine\ORM\Mapping as ORM;
 
-/**
- * @ORM\Entity
- */
```
Comment (martin-georgiev) [inline fixtures/MartinGeorgiev/Doctrine/Entity/ContainsIntegers.php:11]: I'd rather not pollute the PR with such style refactoring changes, that are not related to the said ORM support functionality. Can you please revert and open a separate PR with these phpDoc changes?

File: composer.json [hunk 2]
```diff
@@ -57,8 +56,7 @@
         "symfony/cache": "^6.4||^7.0"
     },
     "conflict": {
-        "doctrine/lexer": "<1.0||>=3.0",
-        "doctrine/orm": "<2.8||>=3.0"
```
Comment (martin-georgiev) [inline composer.json:61]: It will be good to enrich the CI steps with an updated [matrix](https://github.com/martin-georgiev/postgresql-for-doctrine/blob/main/.github/workflows/ci.yml#L20) that explicitly checks both ORM v2 and ORM v3 will run the test suite without issues. For example, the setup may look like the below pseudocode (I haven't run it so, take it with a pinch of salt):
```shell
#.github/workflows/ci.yml

jobs:
  tests:
    name: "PHP ${{ matrix.php }} + Doctrine ORM ${{ matrix.doctrine-orm }} + Composer ${{ matrix.composer-flags }}"
    runs-on: ubuntu-latest

    strategy:
      fail-fast: false
      matrix:
        include:
          - php: '8.1'
          - php: '8.2'
          - php: '8.3'
            ignore-php-version: true
            calculate-code-coverage: true
          - doctrine-orm: '2.8'
          - doctrine-orm: '3.0'
          - doctrine-orm: 'latest'

    - name: Install Doctrine ORM dependencies
      run: |
        if [ "${{ matrix.doctrine-orm }}" == "2.8" ]; then
          composer require doctrine/orm "~2.8" --prefer-dist --no-interaction --no-progress ${{ matrix.composer-flags }}
        elif [ "${{ matrix.doctrine-orm }}" == "3.0" ]; then
          composer require doctrine/orm "~3.0" --prefer-dist --no-interaction --no-progress ${{ matrix.composer-flags }}
        else
          composer update --prefer-dist --no-interaction --no-progress ${{ matrix.composer-flags }}
        fi

File: src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/BaseFunction.php [hunk 3]
```diff
@@ -43,12 +44,14 @@ protected function addNodeMapping(string $parserMethod): void
 
     public function parse(Parser $parser): void
     {
+        $ormV2 = !\class_exists(TokenType::class);
```
Comment (martin-georgiev) [inline src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/BaseFunction.php:47]: As this expression repeats at several places, wdyt about guaranteeing the expression will always remain consistent by extracting into a new [Utils](https://github.com/martin-georgiev/postgresql-for-doctrine/tree/main/src/MartinGeorgiev/Utils) helper class (that will go away one day)?
```php
namespace MartinGeorgiev\Utils;

class DoctrineOrmVersion {
   public static function isVersion3(): bool {
      return \class_exists(\Doctrine\ORM\Query\TokenType::class);
   }
}
```

**Review by martin-georgiev — CHANGES_REQUESTED** [review]
Thank you for this PR! It certainly is good work that helps bring the library into 2024. I left some minor comments about how it can be further improved.

File: fixtures/MartinGeorgiev/Doctrine/Entity/ContainsIntegers.php (hunk 1 above)
Comment (GwendolenLynch) [reply fixtures/MartinGeorgiev/Doctrine/Entity/ContainsIntegers.php:11]: Totally happy to do another PR, but this is linked to the changes in `Tests\MartinGeorgiev\Doctrine\ORM\Query\AST\Functions\TestCase` as v3 EM doesn't have an annotation driver. See https://github.com/doctrine/orm/blob/3.0.x/UPGRADE.md

What do you think?

File: src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/BaseFunction.php (hunk 3 above)
Comment (GwendolenLynch) [reply src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/BaseFunction.php]: As Doctrine has now added `TokenType` in the unreleased `2.19.x` branch, I've gone for `isPre219()`, but welcome better naming ideas!

**Comment by GwendolenLynch** [comment]
Some additional context for the latest commits.

I started by testing on `main`, and tests _currently_ only pass on `doctrine/orm` 2.14 onwards. 

Based on that, this branch also has its tests passing on 2.14+. So I've updated the constraint to ~2.14,  and `doctrine/annotations` is no longer needed.

Background on ORM v2.14.0:
  - Released December 19 2022
  - [9961: Deprecate EntityManager::create()](https://github.com/doctrine/orm/pull/9961)

File: src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/BaseFunction.php [hunk 4]
```diff
@@ -43,12 +45,14 @@ protected function addNodeMapping(string $parserMethod): void
 
     public function parse(Parser $parser): void
     {
+        $ormV2 = DoctrineOrm::isPre219();
```
Comment (martin-georgiev) [inline src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/BaseFunction.php:48]:
```suggestion
        $shouldUseLexer = DoctrineOrm::isPre219();
```

File: src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/BaseVariadicFunction.php [hunk 5]
```diff
@@ -21,18 +23,19 @@ public function feedParserWithNodes(Parser $parser): void
...
 
         $this->nodes[] = $parser->{$this->commonNodeMapping}();
-        if (!isset($lexer->lookahead['type'])) {
+        if ($lexer->lookahead?->type === null) {
             throw new \RuntimeException('The parser\'s "lookahead" property is not populated with a type');
         }
 
-        $aheadType = $lexer->lookahead['type'];
+        $aheadType = $lexer->lookahead->type;
+        $ormV2 = DoctrineOrm::isPre219();
```
Comment (martin-georgiev) [inline src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/BaseVariadicFunction.php:31]:
```suggestion
        $shouldUseLexer = DoctrineOrm::isPre219();
```

File: src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/Cast.php [hunk 6]
```diff
@@ -28,11 +30,13 @@ class Cast extends FunctionNode
 
     public function parse(Parser $parser): void
     {
-        $parser->match(Lexer::T_IDENTIFIER);
-        $parser->match(Lexer::T_OPEN_PARENTHESIS);
+        $ormV2 = DoctrineOrm::isPre219();
```
Comment (martin-georgiev) [inline src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/Cast.php:33]:
```suggestion
        $shouldUseLexer = DoctrineOrm::isPre219();
```

File: src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/StringAgg.php [hunk 7]
```diff
@@ -35,26 +37,28 @@ protected function customiseFunction(): void
 
     public function parse(Parser $parser): void
     {
+        $ormV2 = DoctrineOrm::isPre219();
```
Comment (martin-georgiev) [inline src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/StringAgg.php:40]:
```suggestion
        $shouldUseLexer = DoctrineOrm::isPre219();
```

**Review by martin-georgiev — COMMENTED** [review]
Looks good! Just a nit on the new variable naming.

**Review by martin-georgiev — APPROVED** [review]

**Comment by martin-georgiev** [comment]
Well done @GwendolenLynch!
Thank you for helping the library move forward.
